core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1505,6 +1505,91 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
     }
   }
 
+  def groupWithin[F2[x] >: F[x]](
+      chunkSize: Int,
+      timeout: FiniteDuration
+  )(implicit F: Temporal[F2]): Stream[F2, Chunk[O]] = {
+
+    case class State[+A](os: Chunk[A], supplyEnded: Boolean) {
+
+      val size: Long = os.size.toLong
+
+      // checking if it's empty to avoid early termination of the stream if the producer is faster than consumers
+      val streamExhausted: Boolean = supplyEnded && os.isEmpty
+
+      def endSupply: State[A] = copy(supplyEnded = true)
+
+      def splitAt(idx: Long): (State[A], Chunk[A]) = {
+        val (flushed, kept) = os.splitAt(idx.toInt)
+        (copy(os = kept), flushed)
+      }
+
+      override def toString =
+        s"State(os = $size, os = $os supplyEnded = $supplyEnded, streamExhausted = $streamExhausted)"
+    }
+
+    object State {
+      def add[A](a: A)(s: State[A]): State[A] = s.copy(os = s.os ++ Chunk.singleton(a))
+    }
+
+    if (timeout.toNanos == 0) chunkN(chunkSize)
+    else
+      Stream.force {
+        for {
+          state <- Ref.of(State[O](os = Chunk.empty, supplyEnded = false))
+          supply <- Semaphore[F2](0)
+        } yield {
+
+          val groupSize = chunkSize.toLong
+
+          val enqueue: F2[Unit] =
+            evalTap(o => state.update(State.add[O](o)) *> supply.release)
```

Comment:
Since you are not using the result after it. 

```suggestion
            foreach(o => state.update(State.add[O](o)) *> supply.release)
```